src/seraphis_mocks/mock_http_client_pool.h

```diff
@@ -0,0 +1,131 @@
+
+//forward declarations
+
+namespace sp
+{
+namespace mocks
+{
+////
+// ClientConnectionPool
+// - wraps a pool of network client connections to enable concurrent requests
+///
+class ClientConnectionPool
+{
+public:
+//constructors
+    ClientConnectionPool(
+        const std::string &daemon_address,
+        const boost::optional<epee::net_utils::http::login> daemon_login = boost::none,
+        const epee::net_utils::ssl_options_t ssl_support = epee::net_utils::ssl_support_t::e_ssl_support_autodetect,
+        const size_t &max_connections = 20):
+            m_daemon_address{daemon_address},
+            m_daemon_login{daemon_login},
+            m_ssl_support(ssl_support),
+            m_max_connections{max_connections}
+    {
+        m_http_client_pool.reserve(max_connections);
+    }
+
+//member functions
+    enum http_mode { JSON, BIN, JSON_RPC };
+
+    /// Use an http client from the pool to make an RPC request to the daemon
+    template <typename COMMAND_TYPE>
+    bool rpc_command(const http_mode &mode, const std::string &command_name, const typename COMMAND_TYPE::request &req, typename COMMAND_TYPE::response &res)
+    {
+        // Acquire an http client from the connection pool
+        size_t http_client_index = 0;
+        epee::net_utils::http::abstract_http_client *http_client = acquire_unused_http_client(
+                m_daemon_address,
+                m_daemon_login,
+                m_ssl_support,
+                http_client_index);
+        auto scope_exit_handler = epee::misc_utils::create_scope_leave_handler([this, http_client_index]{
+            release_http_client(http_client_index);
+        });
+
+        // Do the RPC command
+        LOG_PRINT_L2("Invoking " << command_name << " with http client " << http_client_index);
+        switch (mode)
+        {
+            case JSON: return epee::net_utils::invoke_http_json(command_name, req, res, *http_client);
+            case BIN: return epee::net_utils::invoke_http_bin(command_name, req, res, *http_client);
+            case JSON_RPC: return epee::net_utils::invoke_http_json_rpc("/json_rpc", command_name, req, res, *http_client);
+            default: { MERROR("Unknown http_mode: " << mode); return false; }
+        }
+    }
+
+private:
+    /// If an http client is available, acquires it. If none are available,
+    /// initializes a new http client.
+    epee::net_utils::http::abstract_http_client *acquire_unused_http_client(
+        const std::string &daemon_address,
+        const boost::optional<epee::net_utils::http::login> daemon_login,
+        const epee::net_utils::ssl_options_t ssl_support,
+        size_t &http_client_index);
+
+    /// Make http client available for use again.
+    void release_http_client(size_t http_client_index);
+
+//member variables
+private:
+    const std::string m_daemon_address;
+    const boost::optional<epee::net_utils::http::login> m_daemon_login;
+    const epee::net_utils::ssl_options_t m_ssl_support;
```

Comment:
It would be good to put proxy support here

Reply:
Yep, will do